File: repurposeai/src/validate.py
"""
validate.py
Day 4 morning module: retrospective validation. Run the pipeline on your chosen
disease and check whether it recovers a drug with a KNOWN, literature-documented
repurposing success -- this is your strongest evidence slide, stronger than any
feature list.

For Rheumatoid Arthritis specifically, good validation targets include:
  - JAK inhibitors (baricitinib, tofacitinib) -- baricitinib's RA-to-COVID-19
    repurposing is already cited in your deck (ACTT-2 trial), so recovering
    JAK-inhibitor-class candidates from an RA signature is a strong, on-theme result.
  - Anti-TNF biologics (etanercept, adalimumab) -- standard-of-care for RA;
    your pipeline should rank these highly since they're direct validation
    that the pipeline finds real, effective RA drugs, not noise.

Recovering a KNOWN drug is not a failure of "novelty" -- report it as proof the
method works, then separately highlight your top NOVEL candidates.
"""

from __future__ import annotations
import pandas as pd


# Known reference drugs for validation -- extend this list as you research
# your chosen disease further during Day 1-2.
KNOWN_VALIDATION_SETS = {
    "rheumatoid arthritis": {
        "baricitinib", "tofacitinib", "upadacitinib",       # JAK inhibitors
        "etanercept", "adalimumab", "infliximab",             # anti-TNF
        "tocilizumab", "sarilumab",                            # anti-IL6
        "methotrexate",                                        # conventional DMARD
    },
}


def check_recovery(ranked_candidates: pd.DataFrame, disease_name: str,
                    drug_col: str = "drug", top_k: int = 20) -> dict:
    """
    Check how many known reference drugs for `disease_name` appear in the
    top_k ranked candidates. Returns a summary dict ready to drop into the pitch.
    """
    key = disease_name.lower().strip()
    reference_set = KNOWN_VALIDATION_SETS.get(key)
    if reference_set is None:
        raise KeyError(
            f"No reference validation set defined for '{disease_name}'. "
            f"Add one to KNOWN_VALIDATION_SETS in validate.py. "
            f"Available: {list(KNOWN_VALIDATION_SETS)}"
        )

    top = ranked_candidates.head(top_k)
    top_names = {d.lower().strip() for d in top[drug_col]}
    recovered = reference_set & top_names

    return {
        "disease": disease_name,
        "top_k": top_k,
        "reference_set_size": len(reference_set),
        "recovered_drugs": sorted(recovered),
        "recovered_count": len(recovered),
        "recovery_rate": len(recovered) / len(reference_set) if reference_set else 0.0,
        "novel_candidates": [d for d in top[drug_col] if d.lower().strip() not in reference_set],
    }


def format_validation_statement(result: dict) -> str:
    """Human-readable one-liner for the pitch deck / demo narration."""
    if result["recovered_count"] == 0:
        return (
            f"No known {result['disease']} drugs were recovered in the top "
            f"{result['top_k']} candidates -- investigate signature quality before "
            f"presenting novel candidates as high-confidence."
        )
    recovered = ", ".join(result["recovered_drugs"])
    return (
        f"Blind-ran on {result['disease']}: the pipeline correctly surfaced "
        f"{result['recovered_count']}/{result['reference_set_size']} known effective "
        f"treatments ({recovered}) in the top {result['top_k']} candidates, "
        f"validating the method before surfacing {len(result['novel_candidates'])} "
        f"additional novel repurposing candidates."
    )


if __name__ == "__main__":
    print(__doc__)

File: frontend/src/engine/validate.ts
/**
 * validate.ts
 * Port of repurposeai/src/validate.py + app/validation_helpers.py.
 *
 * Two distinct checks, never conflated in the UI:
 *   - checkRecovery()          real known-drug recovery (needs real drug names)
 *   - syntheticGroundTruth()   planted-signal sanity check on the mock matrix
 */
import type { Candidate, RecoveryResult, SyntheticCheck } from './types'

/** Verbatim from validate.KNOWN_VALIDATION_SETS. */
export const KNOWN_VALIDATION_SETS: Record<string, string[]> = {
  'rheumatoid arthritis': [
    'baricitinib',
    'tofacitinib',
    'upadacitinib', // JAK inhibitors
    'etanercept',
    'adalimumab',
    'infliximab', // anti-TNF
    'tocilizumab',
    'sarilumab', // anti-IL6
    'methotrexate', // conventional DMARD
  ],
}

/** Drug-class grouping used only for display in the validation panel. */
export const DRUG_CLASSES: Record<string, string> = {
  baricitinib: 'JAK inhibitor',
  tofacitinib: 'JAK inhibitor',
  upadacitinib: 'JAK inhibitor',
  etanercept: 'anti-TNF biologic',
  adalimumab: 'anti-TNF biologic',
  infliximab: 'anti-TNF biologic',
  tocilizumab: 'anti-IL6 biologic',
  sarilumab: 'anti-IL6 biologic',
  methotrexate: 'conventional DMARD',
}

export const PLANTED_REVERSAL_PREFIX = 'planted_reversal_drug'
export const PLANTED_REINFORCING_PREFIX = 'planted_reinforcing_drug'

export function knownDrugSet(disease: string): Set<string> {
  return new Set(KNOWN_VALIDATION_SETS[disease.toLowerCase().trim()] ?? [])
}

/** True only if the drug library actually contains real reference drug names. */
export function hasRealValidationSignal(drugs: string[], known: Set<string>): boolean {
  const cols = new Set(drugs.map((d) => d.toLowerCase()))
  for (const d of known) if (cols.has(d)) return true
  return false
}

export function checkRecovery(
  ranked: Candidate[],
  diseaseName: string,
  topK: number,
): RecoveryResult {
  const reference = knownDrugSet(diseaseName)
  const top = ranked.slice(0, topK)
  const topNames = new Set(top.map((c) => c.drug.toLowerCase().trim()))
  const recovered = [...reference].filter((d) => topNames.has(d)).sort()

  return {
    disease: diseaseName,
    topK,
    referenceSetSize: reference.size,
    recoveredDrugs: recovered,
    recoveredCount: recovered.length,
    recoveryRate: reference.size ? recovered.length / reference.size : 0,
    novelCandidates: top
      .map((c) => c.drug)
      .filter((d) => !reference.has(d.toLowerCase().trim())),
  }
}

/**
 * Mock-data sanity check: do the planted reversal drugs land near the top and
 * the planted reinforcing (deliberately bad) drugs stay near the bottom?
 * Returns null when no planted naming convention is present.
 */
export function syntheticGroundTruth(
  ranked: Candidate[],
  allDrugs: string[],
  topK: number,
): SyntheticCheck | null {
  const reversalPlanted = allDrugs.filter((c) => c.startsWith(PLANTED_REVERSAL_PREFIX))
  const reinforcingPlanted = allDrugs.filter((c) => c.startsWith(PLANTED_REINFORCING_PREFIX))
  if (reversalPlanted.length === 0) return null

  const top = new Set(ranked.slice(0, topK).map((c) => c.drug))
  const recovered = reversalPlanted.filter((d) => top.has(d)).sort()

  return {
    recovered,
    recoveredCount: recovered.length,
    plantedTotal: reversalPlanted.length,
    recoveryRate: recovered.length / reversalPlanted.length,
    reinforcingPlanted,
    reinforcingRanks: ranked.filter((c) => reinforcingPlanted.includes(c.drug)),
  }
}

export function formatValidationStatement(r: RecoveryResult): string {
  if (r.recoveredCount === 0) {
    return `No known ${r.disease} drugs were recovered in the top ${r.topK} candidates — investigate signature quality before presenting novel candidates as high-confidence.`
  }
  return `Blind-ran on ${r.disease}: the pipeline surfaced ${r.recoveredCount}/${r.referenceSetSize} known effective treatments (${r.recoveredDrugs.join(', ')}) in the top ${r.topK} candidates, validating the method before surfacing ${r.novelCandidates.length} additional novel repurposing candidates.`
}
